google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,377 @@
+
+  private BigQueryWriteClient client;
+  private String streamName;
+  private StreamWriter streamWriter;
+  private Descriptor descriptor;
+
+  /**
+   * Constructs the JsonStreamWriter
+   *
+   * @param builder The Builder object for the JsonStreamWriter
+   */
+  private JsonStreamWriter(Builder builder)
+      throws Descriptors.DescriptorValidationException, IllegalArgumentException, IOException,
+          InterruptedException {
+    Matcher matcher = streamPattern.matcher(builder.streamName);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException("Invalid stream name: " + builder.streamName);
+    }
+
+    this.streamName = builder.streamName;
+    this.client = builder.client;
+    this.descriptor =
+        BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(builder.tableSchema);
+
+    StreamWriter.Builder streamWriterBuilder;
+    if (this.client == null) {
+      streamWriterBuilder = StreamWriter.newBuilder(builder.streamName);
+    } else {
+      streamWriterBuilder = StreamWriter.newBuilder(builder.streamName, builder.client);
+    }
+    setStreamWriterSettings(
+        streamWriterBuilder,
+        builder.channelProvider,
+        builder.credentialsProvider,
+        builder.batchingSettings,
+        builder.retrySettings,
+        builder.executorProvider,
+        builder.endpoint,
+        builder.onSchemaUpdateRunnable);
+    this.streamWriter = streamWriterBuilder.build();
+  }
+  /**
+   * Writes a JSONArray that contains JSONObjects to the BigQuery table by first converting the JSON
+   * data to protobuf messages, then using StreamWriter's append() to write the data. If there is a
+   * schema update, the OnSchemaUpdateRunnable will be used to determine what actions to perform.
+   *
+   * @param jsonArr The JSON array that contains JSONObjects to be written
+   * @param allowUnknownFields if true, json data can have fields unknown to the BigQuery table.
+   * @return ApiFuture<AppendRowsResponse> returns an AppendRowsResponse message wrapped in an
+   *     ApiFuture
+   */
+  public synchronized ApiFuture<AppendRowsResponse> append(
+      JSONArray jsonArr, boolean allowUnknownFields) {
+    return append(jsonArr, -1, allowUnknownFields);
+  }
+
+  /**
+   * Writes a JSONArray that contains JSONObjects to the BigQuery table by first converting the JSON
+   * data to protobuf messages, then using StreamWriter's append() to write the data. If there is a
+   * schema update, the OnSchemaUpdateRunnable will be used to determine what actions to perform.
+   *
+   * @param jsonArr The JSON array that contains JSONObjects to be written
+   * @param offset Offset for deduplication
+   * @param allowUnknownFields if true, json data can have fields unknown to the BigQuery table.
+   * @return ApiFuture<AppendRowsResponse> returns an AppendRowsResponse message wrapped in an
+   *     ApiFuture
+   */
+  public synchronized ApiFuture<AppendRowsResponse> append(
```

Comment:
Is there a reason for it to be a synchronized method? Multiple append can happen at the same time right?

When running together with schema update, maybe only a reader lock is needed for append and a writer lock for schema update method, so that no append will go between refresh and set update schema.

Reply:
Removed synchronized from method signature. Let StreamWriter handle appending the first writer schema after a refreshAppend() so the first schema sent is always right. This removes the need of refreshAppend() having to deal with calling setUpdateSchema(), which is a function specific to JsonStreamWriter and not StreamWriter.